keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+semantically ready.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define API to mark Pod as schedule-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.status.schedulePaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.status.schedulePaused` to the Pod API. The field is defaulted to `false`.
+
+For Pods carrying `.status.schedulePause=true`, they will be "parked" in scheduler's internal
+unschedulablePods pool, and only gets retried when the field is set to `false`.
+
+Practically, this field can be initialized to `true` by MutatingWebhook, and be flipped to `false`
+by external integrators when certain criteria is met.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an application developer, I'm knowledgeable to know when my Pods are schedule-ready, but I know
+nothing about scheduling internals. It'd be ideal for me to only set/unset the "paused" flag,
+instead of developing scheduler plugins, to control the schedulability of my Pods.
+
+#### Story 2
+
+As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins
+to guard the schedulability of my Pods.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+<<[UNRESOLVED]>>
+
+The feature will be guarded by a feature gate and disabled by default in alpha. When alpha, surely 
+the Pod API field needs to be disabled, but I'm open to hear feedback on whether to disable scheduler's
+ComponentConfig API as well as the extension point logic.
+
+If we disable it, then users cannot use the new extension point at all.
+If we don't disable it, people can still customize their enqueue behaviors, like wire it with a
+customized field (like an annotation key).
+
```

Comment:
The feature gate should probably be only on the Pod api not the new scheduler extension point. I don't think we ever used feature gates while developing and evolving the scheduler framework since it was always thought of as a refactor that has no user visible effects.

Reply:
Yes, that's exactly what I thought of. Will let it soak for a while, and then remove UNRESOLVED to clarify the semantics for extension point.